safehttp/plugins/fetch_metadata/fetch_metadata.go

```diff
@@ -0,0 +1,186 @@
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fetchmetadata
+
+import (
+	"github.com/google/go-safeweb/safehttp"
+	"log"
+	"net/url"
+)
+
+// RequestLogger is a user-provided service for logging Fetch Metadata policy
+// violations
+type RequestLogger interface {
+	Log(ir *safehttp.IncomingRequest)
+}
+
+var (
+	navigationalModes = map[string]bool{
+		"navigate":        true,
+		"nested-navigate": true,
+	}
+	navigationalDest = map[string]bool{
+		"audio":           true,
+		"audioworklet":    true,
+		"document":        true,
+		"empty":           true,
+		"font":            true,
+		"image":           true,
+		"manifest":        true,
+		"paintworklet":    true,
+		"report":          true,
+		"script":          true,
+		"serviceworker":   true,
+		"sharedworker":    true,
+		"style":           true,
+		"track":           true,
+		"video":           true,
+		"worker":          true,
+		"xslt":            true,
+		"nested-document": true,
+	}
+	statePreservingMethods = map[string]bool{
+		"GET":  true,
+		"HEAD": true,
+	}
+)
+
+// Plugin implements Fetch Metadata functionality.
+//
+// See https://www.w3.org/TR/fetch-metadata/ and
+// https://web.dev/fetch-metadata/  for more details.
+type Plugin struct {
+	// NavIsolation indicates whether the Navigation Isolation Policy should
+	// be applied to the request before the Resource Isolation Policy as an
+	// additional layer of checking. This provides a way to mitigate
+	// clickjacking and reflected XSS by rejecting all cross-site navigations
+	// unless targeted to endpoints that are CORS-protected.
+	//
+	// WARNING: This is still an experimental feature and will be disabled by
+	// default.
+	NavIsolation bool
+	// RedirectURL can optionally indicate an URL to which the user can be
+	// redirected in case the Navigation Isolation policy rejects the request.
+	RedirectURL   string
+	reportOnly    bool
+	logger        RequestLogger
+	corsProtected map[string]bool
+}
+
+// NewPlugin creates a new Fetch Metadata plugin in enforce mode that will apply
+// the Resource Isolation Policy by default. The user can provide a set of
+// endpoints that are CORS-protected. Any request targeted to those endpoints
+// will be allowed by default without the policies being applied.
+func NewPlugin(endpoints ...string) *Plugin {
+	m := map[string]bool{}
+	for _, e := range endpoints {
+		m[e] = true
+	}
+	return &Plugin{
+		corsProtected: m,
+	}
+}
+
+func (p *Plugin) resourceIsolationPolicy(r *safehttp.IncomingRequest) bool {
+	h := r.Header
+	method := r.Method()
+	site := h.Get("Sec-Fetch-Site")
+	mode := h.Get("Sec-Fetch-Mode")
+	dest := h.Get("Sec-Fetch-Dest")
```

Comment:
Move all variable declarations directly before they are used and remove all variables that are only used once. Instead just put the function call directly in their place.

Reply:
The way you did it now looks good!